ext/reflection/php_reflection.c

```diff
@@ -2690,6 +2754,116 @@ ZEND_METHOD(reflection_parameter, isVariadic)
 }
 /* }}} */
 
+/* {{{ proto public bool ReflectionTypeHint::isArray()
+   Returns whether parameter MUST be an array */
+ZEND_METHOD(reflection_typehint, isArray)
+{
+	reflection_object *intern;
+	typehint_reference *param;
+
+	if (zend_parse_parameters_none() == FAILURE) {
+		return;
+	}
+	GET_REFLECTION_OBJECT_PTR(param);
+
+	RETVAL_BOOL(param->arg_info->type_hint == IS_ARRAY);
+}
+/* }}} */
+
+/* {{{ proto public bool ReflectionTypeHint::isCallable()
+   Returns whether parameter MUST be callable */
+ZEND_METHOD(reflection_typehint, isCallable)
+{
+	reflection_object *intern;
+	typehint_reference *param;
+
+	if (zend_parse_parameters_none() == FAILURE) {
+		return;
+	}
+	GET_REFLECTION_OBJECT_PTR(param);
+
+	RETVAL_BOOL(param->arg_info->type_hint == IS_CALLABLE);
+}
+/* }}} */
+
+/* {{{ proto public bool ReflectionTypeHint::isNullable()
+  Returns whether parameter MAY be null */
+ZEND_METHOD(reflection_typehint, isNullable)
+{
+	reflection_object *intern;
+	typehint_reference *param;
+
+	if (zend_parse_parameters_none() == FAILURE) {
+		return;
+	}
+	GET_REFLECTION_OBJECT_PTR(param);
+
+	RETVAL_BOOL(param->arg_info->allow_null);
+}
+/* }}} */
+
+/* {{{ proto public bool ReflectionTypeHint::isInstance()
+  Returns whether parameter MUST be a class/interface instance */
+ZEND_METHOD(reflection_typehint, isInstance)
+{
+	reflection_object *intern;
+	typehint_reference *param;
+
+	if (zend_parse_parameters_none() == FAILURE) {
+		return;
+	}
+	GET_REFLECTION_OBJECT_PTR(param);
+
+	RETVAL_BOOL(param->arg_info->type_hint == IS_OBJECT);
+}
+/* }}} */
+
+/* {{{ proto public bool ReflectionTypeHint::isScalar()
+  Returns whether parameter MUST be scalar */
+ZEND_METHOD(reflection_typehint, isScalar)
+{
+	reflection_object *intern;
+	typehint_reference *param;
+
+	if (zend_parse_parameters_none() == FAILURE) {
+		return;
+	}
+	GET_REFLECTION_OBJECT_PTR(param);
+
+	RETVAL_BOOL(param->arg_info->type_hint == IS_STRING ||
+		param->arg_info->type_hint == _IS_BOOL ||
+		param->arg_info->type_hint == IS_LONG ||
+		param->arg_info->type_hint == IS_DOUBLE);
+}
+/* }}} */
+
+/* {{{ proto public string ReflectionTypeHint::__toString()
+   Return the text of the type hint */
+ZEND_METHOD(reflection_typehint, __toString)
+{
+	reflection_object *intern;
+	typehint_reference *param;
+
+	if (zend_parse_parameters_none() == FAILURE) {
+		return;
+	}
+	GET_REFLECTION_OBJECT_PTR(param);
+
+	switch (param->arg_info->type_hint) {
+		case IS_ARRAY:    RETURN_STRINGL("array", sizeof("array") - 1);
+		case IS_CALLABLE: RETURN_STRINGL("callable", sizeof("callable") - 1);
+		case IS_OBJECT:   RETURN_STR(zend_string_copy(param->arg_info->class_name));
```

Comment:
This needs to specially handle typehints on internal functions, see http://lxr.php.net/xref/PHP_TRUNK/ext/reflection/php_reflection.c#650 for an example. Would also be good to have a test for getting type info on an internal function.